plugins/outputs/websocket/websocket.go

```diff
@@ -0,0 +1,216 @@
+	serializer serializers.Serializer
+}
+
+// SetSerializer implements serializers.SerializerOutput.
+func (w *WebSocket) SetSerializer(serializer serializers.Serializer) {
+	w.serializer = serializer
+}
+
+// Description of plugin.
+func (w *WebSocket) Description() string {
+	return "Generic WebSocket output writer."
+}
+
+// SampleConfig returns plugin config sample.
+func (w *WebSocket) SampleConfig() string {
+	return sampleConfig
+}
+
+var errInvalidURL = errors.New("invalid websocket URL")
+
+// Init the output plugin.
+func (w *WebSocket) Init() error {
+	if parsedURL, err := url.Parse(w.URL); err != nil || (parsedURL.Scheme != "ws" && parsedURL.Scheme != "wss") {
+		return fmt.Errorf("%w: \"%s\"", errInvalidURL, w.URL)
+	}
+	return nil
+}
+
+// Connect to the output endpoint.
+func (w *WebSocket) Connect() error {
+	tlsCfg, err := w.ClientConfig.TLSConfig()
+	if err != nil {
+		return fmt.Errorf("error creating TLS config: %v", err)
+	}
+
+	dialer := &ws.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: time.Duration(w.ConnectTimeout),
+		TLSClientConfig:  tlsCfg,
+	}
+
+	headers := http.Header{}
+	for k, v := range w.Headers {
+		headers.Set(k, v)
+	}
+
+	conn, resp, err := dialer.Dial(w.URL, headers)
+	if err != nil {
+		return fmt.Errorf("error dial: %v", err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		return fmt.Errorf("wrong status code while connecting to server: %d", resp.StatusCode)
+	}
+
+	w.conn = conn
+	go w.read(conn)
+
+	return nil
+}
+
+func (w *WebSocket) read(conn *ws.Conn) {
+	defer func() { _ = conn.Close() }()
+	if w.ReadTimeout > 0 {
+		if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+			return
+		}
+		conn.SetPingHandler(func(string) error {
+			err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout)))
+			if err != nil {
+				return err
+			}
+			return conn.WriteControl(ws.PongMessage, nil, time.Now().Add(time.Duration(w.WriteTimeout)))
+		})
+	}
+	for {
+		// Need to read a connection (to properly process pings from a server).
+		_, _, err := conn.ReadMessage()
+		if err != nil {
+			if ws.IsUnexpectedCloseError(err, ws.CloseGoingAway, ws.CloseAbnormalClosure) {
+				w.Logger.Errorf("error reading websocket connection: %v", err)
+			}
+			return
+		}
+		if w.ReadTimeout > 0 {
+			if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+				return
+			}
+		}
+	}
+}
+
+// Write writes the given metrics to the destination. Not thread-safe.
+func (w *WebSocket) Write(metrics []telegraf.Metric) error {
+	if w.conn == nil {
+		// Previous write failed with error and ws conn was closed.
+		if err := w.Connect(); err != nil {
+			return err
+		}
+	}
+
+	messageData, err := w.serializer.SerializeBatch(metrics)
+	if err != nil {
+		return err
+	}
+
+	messageType := ws.BinaryMessage
+	if w.UseTextFrames {
+		messageType = ws.TextMessage
+	}
```

Comment:
This can be moved to the `Init()` function I think.

Reply:
That just a frame type detection based on config allocated on stack, moving it to WebSocket struct will only make code harder to follow in my opinion.

Reply:
TBH I like it more if "one-time things" are not in here, as it clutters the code. But if @ssoroka aggrees we can leave it here.